### DISCUSSION TITLE
Not-found header files despite the `compile-commands.json`-equivalent command succeeding fine

### DISCUSSION THREAD

**Discussion by metaleap** [clangd/clangd#1853] [Q&A] [answered]
Probably a newbie question, but cannot figure out what on earth I might be missing. So basically imagine, without knowing all the project's nuts and bolts (no need IMHO), this simple `compile_commands.json`:

```json
[
    {
        "dir": "/home/_/c/c/mo",
        "file": "mo2d/main.cpp",
        "arguments": [ "clang++", "-c", "-g", 
                                  "-Ilibdeps/SFML/include/", 
                                  "mo2d/main.cpp", 
                                  "-o", "/home/_/.local/bin/mo2d_dbg", 
                                  "-O0" ]
    }
]
```

- The `-I` path is correct,
- both absolute and relative paths bring the same problem, 
- with or without trailing slash,
- and running the command in the terminal raises not a single error:

```
$ rm ~/tmp/mo2d_dbg
$ clang++ -c -g -Ilibdeps/SFML/include/ mo2d/main.cpp -o /home/_/tmp/mo2d_dbg -O0
$ file ~/tmp/mo2d_dbg
/home/_/tmp/mo2d_dbg: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), with debug_info, not stripped
```

**The actual problem?** The LSP-clienting VSCode extension triggers diagnostics / "Problems" like `'SFML/Graphics.hpp' file not found    clang(pp_file_not_found)`. (And about all the lines using the includes, of course.)

The includes are from a shared library (loaded at program startup, not statically linked), I might add. But I was reasonably confident that link-related stuff don't matter in a _compile_-commands DB.

What else is needed in the setup?

**Comment by i-ky** [comment, marked as answer]
According to [format specification](https://clang.llvm.org/docs/JSONCompilationDatabase.html) it's `"directory"`, not `"dir"`.

**Reply by metaleap** [reply]
Good catch, thanks! Did the trick too. Don't know where I picked up `dir` then, must have Googled into by-now-outdated infos there. Merci @i-ky !

**Reply by HighCommander4** [reply]
I just wanted to add that I don't recommend trying to write `compile_commands.json` manually.

It's better to generate it by running a tool like [Bear](https://github.com/rizsotto/Bear) wrapping the build command (e.g. if using `make`, running `bear -- make`).

The advantages of this are:

 1. You don't need to worry about getting the format of `compile_commands.json` right
 2. If you need to add a new flag, you only need to add it in one place (e.g. your build script), and not separately to `compile_commands.json`
